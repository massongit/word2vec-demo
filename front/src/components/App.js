import React from "react"
import PropTypes from "prop-types"
import Header from "./Header"
import InputPanel from "./InputPanel"
import MethodPanel from "./MethodPanel"
import DescriptionPanel from "../containers/DescriptionPanel"
import OutputPanel from "./OutputPanel"
import {Grid} from "react-bootstrap"

/**
 * similarのPropTypes
 * @type {{similar: shim}}
 */
export const propTypesSimilar = {
    similar: PropTypes.arrayOf(PropTypes.shape({
        word: PropTypes.string.isRequired,
        cosine: PropTypes.number.isRequired
    }).isRequired).isRequired
}

/**
 * ルート要素
 */
class App extends React.Component {
    static propTypes = {
        ...propTypesSimilar,
        method: PropTypes.string
    }

    render() {
        return (
            <Grid>
                <Header/>
                <DescriptionPanel/>
                <MethodPanel/>
                {
                    (() => {
                        if (this.props.method) {
                            return (<InputPanel/>)
                        }
                    })()
                }
                {
                    (() => {
                        // 分散表現による計算が行われていない場合には表示しない
                        if (0 < this.props.similar.length) {
                            return (<OutputPanel/>)
                        }
                    })()
                }
            </Grid>
        )
    }
}

export default App
