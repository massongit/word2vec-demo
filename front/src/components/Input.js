import React from "react"
import fetch from "node-fetch"
import PropTypes from "prop-types"
import {Button, Form, FormControl, InputGroup} from "react-bootstrap"
import {FormattedMessage, intlShape} from "react-intl"
import {showSimilarWords} from "../actions"
import {propTypesPN} from "./KeyWords"

/**
 * 入力部
 */
class Input extends React.Component {
    static propTypes = {
        dispatch: PropTypes.func.isRequired,
        keywords: PropTypes.shape(propTypesPN).isRequired,
        intl: intlShape.isRequired
    }

    constructor(props) {
        super(props)
        this.onSubmit = this.onSubmit.bind(this)
    }

    /**
     * 入力されたキーワードをpositiveな単語とnegativeな単語に分類する
     * @returns {{positive: Array, negative: Array}} APIへの入力用のキーワード
     */
    classifyKeyWords() {
        // キーワード
        const keywords = {
            positive: [],
            negative: []
        }

        for (const word of this.input.value.replace("　", " ").trim().split(" ")) {
            if (word.startsWith("-")) {
                keywords.negative.push(word.slice(1))
            } else if (word.startsWith("+")) {
                keywords.positive.push(word.slice(1))
            } else if (word) {
                keywords.positive.push(word)
            }
        }

        return keywords
    }

    /**
     * 入力されたキーワードを元にAPIへの入力用のキーワードを作成する
     * @returns {{positive: Array, negative: Array}} APIへの入力用のキーワード
     */
    makeKeyWords() {
        // キーワード
        const keywords = this.classifyKeyWords()

        // キーワードのりストをソート
        for (const k in keywords) {
            keywords[k].sort()
        }

        return keywords
    }

    /**
     * onSubmitイベント
     * @param ev イベント
     */
    async onSubmit(ev) {
        // サーバーへのSubmitが行われないようにする
        ev.preventDefault()

        // キーワード
        const keywords = this.makeKeyWords()

        // 以前の入力文とは異なる文章が入力されたとき、Word2Vecによる計算を行い、Word2Vecによる計算結果の表示Actionをdispatch
        if ((0 < keywords.positive.length || 0 < keywords.negative.length) && JSON.stringify(keywords) !== JSON.stringify(this.props.keywords)) {
            try {
                this.props.dispatch(showSimilarWords({
                    keywords,
                    words: await this.callWord2Vec(keywords)
                }))
            } catch (er) {
                alert(this.props.intl.formatMessage(
                    {
                        id: "errorMessage.predict"
                    },
                    {
                        message: er.message
                    }
                ))
            }
        }
    }

    /**
     * Word2VecのAPIを呼び出す
     * @param query {object} クエリ
     * @returns {Promise<*>} 計算結果
     */
    async callWord2Vec(query) {
        return await (await fetch(
            "/word2vec",
            {
                method: "POST",
                headers: {
                    "Accept": "application/json",
                    "Content-Type": "application/json"
                },
                body: JSON.stringify(query)
            }
        ).then(res => {
            if (res.ok) {
                return res
            } else {
                throw new Error(res.statusText)
            }
        })).json()
    }

    render() {
        return (
            <Form
                onSubmit={this.onSubmit}
            >
                <InputGroup>
                    <FormControl
                        defaultValue="王 -男 女"
                        inputRef={n => {
                            this.input = n
                        }}
                    />
                    <InputGroup.Button>
                        <Button type="submit">
                            <FormattedMessage id="calculate"/>
                        </Button>
                    </InputGroup.Button>
                </InputGroup>
            </Form>
        )
    }
}

export default Input
