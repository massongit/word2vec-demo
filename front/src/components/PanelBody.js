import React from "react"
import PropTypes from "prop-types"
import {Panel} from "react-bootstrap"
import {intlShape} from "react-intl"

/**
 * PanelのBody
 */
class PanelBody extends React.Component {
    static propTypes = {
        messageId: PropTypes.string,
        additionalDescription: PropTypes.node,
        children: PropTypes.node,
        intl: intlShape.isRequired
    }

    /**
     * 説明部を描画する
     * @returns {*} 説明部
     */
    renderDescription() {
        if (this.props.messageId) {
            const description = [
                this.props.intl.formatMessage({
                    id: this.props.messageId
                }),
                this.props.additionalDescription
            ]

            if (this.props.children) {
                return (
                    <p>
                        {description}
                    </p>
                )
            } else {
                return description
            }
        }
    }

    render() {
        return (
            <Panel.Body>
                {this.renderDescription()}
                {this.props.children}
            </Panel.Body>
        )
    }
}

export default PanelBody
