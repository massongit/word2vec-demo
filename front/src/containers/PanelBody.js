import PanelBody from "../components/PanelBody"
import {injectIntl} from "react-intl"

/**
 * PanelのBodyのContainer
 */
export default injectIntl(PanelBody)
