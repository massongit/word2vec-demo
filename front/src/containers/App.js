import App from "../components/App"
import {connect} from "react-redux"

const mapStateToProps = state => ({
    similar: state.showSimilarWords.words.similar,
    method: state.setMethod.method
})

/**
 * ルート要素のContainer
 */
export default connect(mapStateToProps)(App)
