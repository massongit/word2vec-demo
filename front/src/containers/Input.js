import Input from "../components/Input"
import {connect} from "react-redux"
import {injectIntl} from "react-intl"

const mapStateToProps = state => ({
    keywords: state.showSimilarWords.keywords,
    loading: state.loading.loading,
    prevMethod: state.showSimilarWords.method,
    method: state.setMethod.method
})

/**
 * 入力部のContainer
 */
export default injectIntl(connect(mapStateToProps)(Input))
