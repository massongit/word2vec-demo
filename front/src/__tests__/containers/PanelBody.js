import React from "react"
import PanelBody from "../../containers/PanelBody"
import Input from "../../containers/Input"
import rootReducer from "../../reducers/index"
import {Panel} from "react-bootstrap"
import {loadTranslation, mountWithIntl, shallowWithIntl} from "enzyme-react-intl"
import {createStore} from "redux"
import {functions2} from "./App"

loadTranslation("./src/translations/ja.json")

let store

describe("containers/PanelBody", () => {
    beforeEach(() => {
        store = createStore(rootReducer)
        for (const f of functions2) {
            if (f) {
                f(store)
            }
        }
    })

    it("Componentが正しく配置されている", () => {
        const panelBodyComponent = shallowWithIntl(
            <PanelBody messageId="description.input">
                <Input
                    store={store}
                />
            </PanelBody>
        ).dive()
        expect(panelBodyComponent).toMatchSnapshot()
    })

    it("Panel.Bodyの直下に子要素が入っている", () => {
        const panelBodyComponent = mountWithIntl(
            <PanelBody messageId="description.input">
                <Input
                    store={store}
                />
            </PanelBody>
        )
        expect(panelBodyComponent.find(Panel.Body).children().contains(Input)).toBeTruthy()
    })
})
