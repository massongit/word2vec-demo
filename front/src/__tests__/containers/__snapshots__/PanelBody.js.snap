// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/PanelBody Componentが正しく配置されている 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <PanelBody
    intl={
        Object {
            "defaultFormats": Object {},
            "defaultLocale": "en",
            "formatDate": [Function],
            "formatHTMLMessage": [Function],
            "formatMessage": [Function],
            "formatNumber": [Function],
            "formatPlural": [Function],
            "formatRelative": [Function],
            "formatTime": [Function],
            "formats": Object {},
            "formatters": Object {
              "getDateTimeFormat": [Function],
              "getMessageFormat": [Function],
              "getNumberFormat": [Function],
              "getPluralFormat": [Function],
              "getRelativeFormat": [Function],
            },
            "locale": "en",
            "messages": Object {
              "calculate": "計算",
              "description.input": "日本語の単語を空白で区切って入力し、計算ボタンをクリックしてください。",
              "description.input.supplement.1": "引き算を行う場合には、単語の先頭に『-』を付けます。",
              "description.input.supplement.2": "足し算を行う場合には、単語の先頭に『+』を付けることができます (基本的には付けなくても問題ありませんが、単語が『+』や『-』で始まる場合には必ず付けてください)。",
              "description.word2Vec": "単語をベクトル表現するための技術です。これを利用して、単語の足し算や引き算を行うことができます。",
              "errorMessage.default": "エラーが発生しました: {message}",
              "errorMessage.predict": "計算中にエラーが発生しました: {message}",
              "header.candidate": "候補",
              "header.cosine": "コサイン類似度 (-1 〜 1)",
              "header.keywords": "<キーワード>",
              "header.word": "単語",
              "title.description": "概要",
              "title.input": "入力",
              "title.result": "結果",
              "unknown": "(予測不能)",
            },
            "now": [Function],
            "textComponent": "span",
          }
    }
    messageId="description.input"
>
    <InjectIntl(Connect(Input)) />
</PanelBody>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "bsClass": "panel",
      "children": Array [
        <p>
          日本語の単語を空白で区切って入力し、計算ボタンをクリックしてください。
</p>,
        <InjectIntl(Connect(Input)) />,
      ],
      "collapsible": false,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "日本語の単語を空白で区切って入力し、計算ボタンをクリックしてください。",
            undefined,
          ],
        },
        "ref": null,
        "rendered": Array [
          "日本語の単語を空白で区切って入力し、計算ボタンをクリックしてください。",
          undefined,
        ],
        "type": "p",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "bsClass": "panel",
        "children": Array [
          <p>
            日本語の単語を空白で区切って入力し、計算ボタンをクリックしてください。
</p>,
          <InjectIntl(Connect(Input)) />,
        ],
        "collapsible": false,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "日本語の単語を空白で区切って入力し、計算ボタンをクリックしてください。",
              undefined,
            ],
          },
          "ref": null,
          "rendered": Array [
            "日本語の単語を空白で区切って入力し、計算ボタンをクリックしてください。",
            undefined,
          ],
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "context": Object {
      "intl": Object {
        "defaultFormats": Object {},
        "defaultLocale": "en",
        "formatDate": [Function],
        "formatHTMLMessage": [Function],
        "formatMessage": [Function],
        "formatNumber": [Function],
        "formatPlural": [Function],
        "formatRelative": [Function],
        "formatTime": [Function],
        "formats": Object {},
        "formatters": Object {
          "getDateTimeFormat": [Function],
          "getMessageFormat": [Function],
          "getNumberFormat": [Function],
          "getPluralFormat": [Function],
          "getRelativeFormat": [Function],
        },
        "locale": "en",
        "messages": Object {
          "calculate": "計算",
          "description.input": "日本語の単語を空白で区切って入力し、計算ボタンをクリックしてください。",
          "description.input.supplement.1": "引き算を行う場合には、単語の先頭に『-』を付けます。",
          "description.input.supplement.2": "足し算を行う場合には、単語の先頭に『+』を付けることができます (基本的には付けなくても問題ありませんが、単語が『+』や『-』で始まる場合には必ず付けてください)。",
          "description.word2Vec": "単語をベクトル表現するための技術です。これを利用して、単語の足し算や引き算を行うことができます。",
          "errorMessage.default": "エラーが発生しました: {message}",
          "errorMessage.predict": "計算中にエラーが発生しました: {message}",
          "header.candidate": "候補",
          "header.cosine": "コサイン類似度 (-1 〜 1)",
          "header.keywords": "<キーワード>",
          "header.word": "単語",
          "title.description": "概要",
          "title.input": "入力",
          "title.result": "結果",
          "unknown": "(予測不能)",
        },
        "now": [Function],
        "textComponent": "span",
      },
    },
  },
}
`;
